### DISCUSSION TITLE
Advice on how to import posting_date from Saltedge

### DISCUSSION THREAD

**Discussion by scottdk** [firefly-iii/firefly-iii#8231] [Firefly III]
EDIT:

I found the `journal_meta` table that now contains the journal date fields. 
---

EDIT:  I have made the changes to pull `posting_date` from Saltedge to populate `process_date`. https://github.com/scottdk/firefly-data-importer/compare/firefly-iii:data-importer:main...main. Unfortunately I wasn't able to create a pull request with this change. (And why `process_date` is still null in my mariadb queries to `transactions_journals` but I see it fine in the app is my next mystery. SOLVED. See edit)

---

My credit card's statements use the transaction's `posting_date` whereas the Data Importer is pulling in `made_on` from Saltedge. I prefer to have `made_on` as I can see when I actually made the transaction, but it does complicate reconciliation with the statements.

Additionally, when I import transactions from the bank's CSV export, it uses the Posting date, so importing transaction earlier than those contained in Saltedge will have the `posting_date` in Firefly's `date` field in the transaction journal, while Saltedge imports have `made_on`.

I would like to either:

1. Request a change to allow importing of the `posting_date` field from Saltedge to `book_date` or `process_date` in the transaction journal (whichever is more meaningful) or guidance on how to do this myself.  (I'm partway there, just looking for how to upload `process_date` to the Firefly DB.)
_2. Alternatively, I could kludge modifying my installation to import `posting_date` as the `madeOn` date This isn't ideal though as data is lost. I'm sure once I find my way around the code, I could implement scenario **1** above._ EDIT: Reduntant. 

**1** above is ideal, and I can probably do it myself and make a pull request. (Rusty with git, but I could in the very least send you my code). 

Ideally I could also choose to 1. sort the `Transaction history overview (audit)` report by posting_date and 2. see which transactions are already reconciled in that report, although I have a spreadsheet that makes an ODBC query to the database which I use to reconcile transactions against my bank's csv downloads. I'll make a separate request to show Reconciled status and to be able to sort by Processing Date in the audit report.

I also plan on making a git project to share my reconciliation spreadsheet once it becomes a bit more generic, or leave it as is for people who know what they're doing with Excel and ODBC...  _EDIT: If I can work out Process Date is still null in mydb queries. I have only one instance running..._

### JSON from Saltedge:

`    {

        "id": xxxxxxxxxxxxxxx8640,
        "account_id": xxxxxxxxxxxxxxx7188,
        "duplicated": false,
        "mode": "normal",
        "status": "posted",
        "made_on": "2023-11-26",
        "amount": -8,
        "currency_code": "AUD",
        "description": "MELBOURNE AIRPORT TULLAMARINE",
        "category": "service_fee",
        "similar_pending_ids": null,
        "extra": {
            "posting_date": "2023-11-29",
            "account_balance_snapshot": 0,
            "categorization_confidence": 0.6
        },
        "created_at": "2023-12-01T03:57:35Z",
        "updated_at": "2023-12-01T03:57:35Z"
    },`

### transaction_journals entry
![image](https://github.com/firefly-iii/firefly-iii/assets/11282782/5b93a0c6-dc0c-4959-967e-0f21acf594c1)

### Modifying the Data Importer myself

EDIT: I've made the change and it's working on my fork.

_I'm not a PHP programmer, but I can find my way around docker containers in VSCODE._

_I can see that_ `postingDate` _is gathered in_ `./app/Services/Spectre/Model/TransactionExtra.php`

_Would uploading this to Firefly just require replicating the code for _`madeOn` _in
`./app/Services/Spectre/Conversion/Routine/GenerateTransactions.php`_,  and_
`./app/Services/Spectre/Conversion/Routine/TransactionProcessor.php` _to perform the same for `postingDate`_? (Minus the transaction filtering, of course)._

_I will_ **already DID** _spin up another instance and try it, but hoping to save some time with some tips before I dive in and get my hands dirty._

**Comment by scottdk** [comment]
Coment Post-edits: Thanks to this problem, I've now dived into PHP for the first time, plus created my first ever docker build. It's based on a fork from the original repo, now I found it's Dockerfile on Azure . 

If this change gets implemented, I'll point my Dockerfile back to the official repository instead of my own fork.

**Comment by JC5** [comment]
Hey Scott, thanks for your patience. I pushed your issue to the bottom of my todo list to pick up various other things. I'll check this out over the next few days. I'm hesitant to change these kinds of things because when I change it for one user, you know it's going to mess up other users. 

Having said that, if the change makes sense it'll be part of the main repo soon. Thanks again!

**Reply by scottdk** [reply]
Yes, I'm using the change in my local install, and everything is hummiong along fine. It's quite a small change really. Just adding one more field to those available in Saltedge that wasn't being imported previously. I'm happy just running it locally, but it would be nice to not have to run my own build.

**Comment by JC5** [comment]
> (And why process_date is still null in my mariadb queries to transactions_journals but I see it fine in the app is my next mystery)

This column is unused nowadays, because it's not very flexible (there are now about 8 dates you can associate with a transaction). Check out the table `journal_meta`, that's where you'll find it.

**Reply by scottdk** [reply]
Ah, yes. I found that table a little while after. I have my ODBC queries running into excel perfectly now, including process_date. Reconciliation is so much easier.

**Comment by JC5** [comment]
Alright, I finally spent more time on this.

- The fake banks provided by Spectre do not include `time`, `post_date` or `post_time`, unfortunately. So I'm still a little limited in my options.
- I've added all three fields to the notes (if the fields are not NULL) and to the logs. That should show me what these fields tend to contain (the API docs aren't very clear on this).
- In the next release, you may see duplicate imports because of these changes.

I hope that this leads to a better import for everybody. Thanks for your patience again, and check out the next release of the data importer and let me know what you think.
